### PR TITLE
fix: conversation performance issues

### PR DESCRIPTION
**Pull request by adrien-coye**
## Abstract

Pushing forward with rendering optimisations of `MessageView` in the context of a list.

## Detail

- Performance issues with preprocessing one mail is multiplied
  - Gains on `MessageView` helps here, but not enough.
- All collapsed mails are processed -> now lazy.
- Cancel preprocess task on collapse.
- Cache processed mail.
- Enable downloading of images in parallel.

## TODO
- Use core lib here https://github.com/Infomaniak/ios-core/pull/42

### CODE REVIEW

File: Mail/Views/Thread/MessageView.swift [hunk 1]
```diff
@@ -26,13 +26,121 @@ import RealmSwift
 import Shimmer
 import SwiftUI
 
+// TODO: Move to core
```
Comment (adrien-coye) [inline Mail/Views/Thread/MessageView.swift:29]: TODO: move to core before merge

File: Mail/Views/Thread/MessageView.swift [hunk 2]
```diff
@@ -26,13 +26,121 @@ import RealmSwift
...
+extension Task {
+    @discardableResult
+    func finish() async -> Result<Success, Failure> {
+        await result
+    }
+}
+
+// TODO: Move to core
```
Comment (adrien-coye) [inline Mail/Views/Thread/MessageView.swift:37]: TODO: move to core before merge

File: Mail/Views/Thread/MessageView.swift [hunk 3]
```diff
@@ -26,13 +26,121 @@ import RealmSwift
...
+    }
+
+    /// The accumulated ordered result at the time of calling. Nil values are removed.
+    /// - Returns: The ordered result at the time of calling. Nil values are removed.
+    public var compactAccumulation: [T] {
+        return buffer.compactMap { $0 }
+    }
+}
+
+// TODO: Move to core
```
Comment (adrien-coye) [inline Mail/Views/Thread/MessageView.swift:80]: TODO: move to core before merge

File: Mail/Views/Thread/MessageView+Preprocessing.swift [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        guard let messageBody = message.body else {
+            return
+        }
+
+        presentableBody.body = messageBody.detached()
+        let bodyValue = messageBody.value ?? ""
+
+        // Heuristic to give up on mail too large for "perfect" preprocessing.
+        // 1 Meg looks like a fine threshold
+        guard bodyValue.lengthOfBytes(using: String.Encoding.utf8) < 1_000_000 else {
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply Mail/Views/Thread/MessageView+Preprocessing.swift:66]: sure thing 👍

File: Mail/Views/Thread/MessageView+Preprocessing.swift [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                    of: "cid:\(contentId)",
+                    with: "data:\(attachment.mimeType);base64,\(data.base64EncodedString())"
+                )
+            }
+
+            // Mutate DOM
+            await mutate(body: mailBody, compactBody: compactBody)
+
+            // Delay between each chunk processing just enough, so the user feels the UI is responsive.
+            try await Task.sleep(nanoseconds: 4_000_000_000)
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply Mail/Views/Thread/MessageView+Preprocessing.swift:145]: yup

**Review by valentinperignon — APPROVED** [review]

**Pull request merged**